src/BuildLoop.cpp

```diff
@@ -150,27 +160,237 @@ static void SignalMainThreadToStartCleaningUp(BuildQueue *queue)
+{
+    if (!queue->m_Config.m_AttemptCacheReads && !queue->m_Config.m_AttemptCacheWrites)
+        return false;
+    return 0 != (node->m_DagNode->m_Flags & Frozen::DagNode::kFlagCacheableByLeafInputs);
+}
+
+static void AttemptCacheWrite(BuildQueue* queue, ThreadState* thread_state, RuntimeNode* node)
+{
+    CheckDoesNotHaveLock(&queue->m_Lock);
+
+    uint64_t time_exec_started = TimerGet();
+
+    char path[kMaxPathLength];
+    char digestString[kDigestStringSize];
+    DigestToString(digestString, node->m_CurrentLeafInputSignature->digest);
+    snprintf(path, sizeof(path), "%s", digestString);
+    FILE *sig = fopen(path, "w");
+    if (sig == NULL)
+    {
+        printf("Failed to open file for signature ingredient writing. Skipping CacheWrite.\n");
+        return;
+    }
+
+    //we already calculated the leaf input signature before, but we'll do it again because now we want to have the ingredient stream written out to disk.
+    CalculateLeafInputSignature(queue, node->m_DagNode, node, &thread_state->m_ScratchAlloc, thread_state->m_ProfilerThreadId, sig);
+
+    fflush(sig);
+    fclose(sig);
+
+    auto writeResult = CacheClient::AttemptWrite(queue->m_Config.m_Dag, node->m_DagNode, node->m_CurrentLeafInputSignature->digest, queue->m_Config.m_StatCache, &queue->m_Lock, thread_state, path);
+    remove(path);
+
+    uint64_t now = TimerGet();
+    double duration = TimerDiffSeconds(time_exec_started, now);
+
+    MutexLock(&queue->m_Lock);
+    PrintMessage(writeResult == CacheResult::Success ? MessageStatusLevel::Success : MessageStatusLevel::Warning, duration, "%s [CacheWrite %s]", node->m_DagNode->m_Annotation.Get(), digestString);
+    MutexUnlock(&queue->m_Lock);
+}
+
+static NodeBuildResult::Enum ExecuteNode(BuildQueue* queue, RuntimeNode* node, Mutex *queue_lock, ThreadState* thread_state, StatCache* stat_cache, const Frozen::DagDerived* dagDerived)
+{
+    CheckDoesNotHaveLock(&queue->m_Lock);
+
+    if (IsNodeCacheableByLeafInputsAndCachingEnabled(queue,node))
+    {
+        CHECK(node->m_CurrentLeafInputSignature != nullptr);
+        bool stillTheSame = node->m_BuiltNode && node->m_BuiltNode->m_WasBuiltSuccessfully && node->m_BuiltNode->m_LeafInputSignature == node->m_CurrentLeafInputSignature->digest;
+
+        if (!stillTheSame)
+            if (!VerifyAllVersionedFilesIncludedByGeneratedHeaderFilesWereAlreadyPartOfTheLeafInputs(queue, thread_state, node, dagDerived))
+                return NodeBuildResult::kRanFailed;
+    }
+
+    bool haveToRunAction = CheckInputSignatureToSeeNodeNeedsExecuting(queue, thread_state, node);
+    if (!haveToRunAction)
+        return NodeBuildResult::kUpToDate;
+
+    NodeBuildResult::Enum runActionResult = RunAction(queue, thread_state, node, queue_lock);
+
+    if (runActionResult == NodeBuildResult::kRanSuccesfully && queue->m_Config.m_AttemptCacheWrites && IsNodeCacheableByLeafInputsAndCachingEnabled(queue,node))
+        AttemptCacheWrite(queue,thread_state,node);
```

Comment:
Is it worth leaving a TODO comment, to consider queueing the cache write (in such a way that tundra doesn't exit while the queue is still being processed)?

Reply:
I think it may make sense to do that as an optimization - if we find that performance is an issue here.

But: I'm not a big fan of TODOs like that in the code base. They have a history of sticking around for ages, often outlasting the problem they are indicating, and in my experience do very little to actually get you to fixing that problem - because whenever you run over that TODO in code, it's usually because you are busy working on something else with that code - which you don't want to stop doing in order to work on this other non-trivial problem the TODO is talking about. So, I think such TODOs are better served living in a TODO list you use to pick up tasks when you're not sure what to work on next.

I personally think TODOs in code comments are only useful in the form:
```
// TODO: (Note to self: Fix this before landing the code. DO NOT APPROVE!)
```

Reply:
OK. (I personally use "FIXME" for things that should not land, and "TODO" for things that can land but someone should probably eventually do something about.)